core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/stopstartds/StopStartDownscaleFlowService.java

```diff
@@ -0,0 +1,116 @@
+import com.sequenceiq.cloudbreak.api.endpoint.v4.common.Status;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.base.InstanceStatus;
+import com.sequenceiq.cloudbreak.cloud.model.CloudVmInstanceStatus;
+import com.sequenceiq.cloudbreak.core.flow2.stack.CloudbreakFlowMessageService;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceMetaData;
+import com.sequenceiq.cloudbreak.service.StackUpdater;
+import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
+import com.sequenceiq.cloudbreak.service.stack.InstanceMetaDataService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+
+@Component
+public class StopStartDownscaleFlowService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StopStartDownscaleFlowService.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private CloudbreakFlowMessageService flowMessageService;
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    public void clusterDownscaleStarted(long stackId, String hostGroupName, Set<Long> privateIds) {
+        clusterService.updateClusterStatusByStackId(stackId, DetailedStackStatus.DOWNSCALE_BY_STOP_IN_PROGRESS);
+        flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_INIT, hostGroupName);
+
+        // TODO CB-15153: Change the message once an adjustment based downscale is supported.
+        LOGGER.debug("stopstart scaling Decommissioning from group: {}, privateIds: [{}]", hostGroupName, privateIds);
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        List<String> decomissionedHostNames = stackService.getHostNamesForPrivateIds(stack.getInstanceMetaDataAsList(), privateIds);
+        LOGGER.debug("stopstart scaling Decommissioning from group: {}, hostnames: [{}]", hostGroupName, decomissionedHostNames);
+        flowMessageService.fireInstanceGroupEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), hostGroupName,
+                CLUSTER_SCALING_STOPSTART_DOWNSCALE_STARTING,
+                String.valueOf(decomissionedHostNames.size()), hostGroupName, String.join(",", decomissionedHostNames));
+    }
+
+    public void logCouldNotDecommission(long stackId, List<String> notDecommissionedFqdns) {
+        // TODO CB-14929: CB-15418 This needs to be an orange message (i.e. not success, not failure). Need to figure out how the UI
+        //  processes these and applies icons.
+        flowMessageService.fireEventAndLog(stackId, UPDATE_FAILED.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_COULDNOTDECOMMISSION,
+                String.valueOf(notDecommissionedFqdns.size()), notDecommissionedFqdns.stream().collect(Collectors.joining(", ")));
```

Comment:
String.join

Reply:
Don't understand this.

Reply:
Please use String.join method like you do other places in this class instead of .stream().collect(Collectors.joining(", ")).

Reply:
Why?

Reply:
Using this when streams are otherwise not required. Would still like to understand the reasoning for the request though.